gapir/cc/memory_allocator.h

```diff
@@ -0,0 +1,144 @@
+                                          : _backing->second->first) == rhs;
+    }
+    bool operator!=(unsigned char* rhs) const {
+      return (_backing == _dummyMap.end() ? nullptr
+                                          : _backing->second->first) != rhs;
+    }
+    unsigned char& operator*() const { return *_backing->second->first; }
+    unsigned char& operator[](size_t n) const {
+      return _backing->second->first[n];
+    }
+
+   private:
+    Handle(std::map<unsigned int,
+                    std::map<unsigned char*, MemoryRegion>::iterator>::iterator
+               backing)
+        : _backing(backing) {}
+    std::map<unsigned int,
+             std::map<unsigned char*, MemoryRegion>::iterator>::iterator
+        _backing;
+
+    static std::map<unsigned int,
+                    std::map<unsigned char*, MemoryRegion>::iterator>
+        _dummyMap;
+
+    friend class MemoryAllocator;
+  };
+
+ public:
+  static std::unique_ptr<MemoryAllocator> create(size_t heapSize);
+
+  MemoryAllocator(size_t heapSize);
+  ~MemoryAllocator();
+
+  Handle allocateStatic(size_t size);
+  Handle allocatePurgable(size_t size, bool allowRelocate = true);
+
+  bool resizeStaticAllocation(const Handle& address, size_t size);
+
+  bool releaseAllocation(const Handle& address);
+
+  bool garbageCollect() { return compactPurgableMemory(); }
+
+  // Get the amount of memory used for different classes of allocation. Warning:
+  // This currently has O(N) cost. If that is causing you trouble go ahead and
+  // maintain an internal record so you can answer this in O(1).
+  size_t getTotalSize() const { return _heapSize; }
+  size_t getTotalStaticDataUsage() const;
+  size_t getTotalPurgableDataUsage() const;
+  size_t getTotalDataUsage() const {
+    return getTotalStaticDataUsage() + getTotalPurgableDataUsage();
+  }
+
+ private:
+  void purgeOrRelocateRange(size_t start, size_t end);
+  bool compactPurgableMemory();
+
+  size_t getClosestStaticData(
+      size_t belowOffset,
+      std::map<unsigned char*, MemoryRegion>::const_iterator*
+          closestStaticIter = nullptr) const;
+
+  Handle registerStaticAllocate(const MemoryRegion& newRegion);
+  Handle registerPurgableAllocate(const MemoryRegion& newRegion);
+
+  void registerResize(const MemoryRegion& resizedRegion);
+
+  void registerRelocate(const MemoryRegion& from, const MemoryRegion& to);
+  void registerPurge(const MemoryRegion& purge);
+
+  void registerStaticRelease(const MemoryRegion& release);
+  void registerPurgableRelease(const MemoryRegion& release);
+
+  size_t _heapSize;
+  unsigned char* _heap;
+
+  size_t _purgableHead;
+
+  std::map<unsigned char*, MemoryRegion> _staticRegionMap;
+  std::map<unsigned char*, MemoryRegion> _purgableRegionMap;
+
+  std::map<unsigned int, std::map<unsigned char*, MemoryRegion>::iterator>
+      _relocationMap;
+  std::map<unsigned char*, unsigned int> _inverseRelocationMap;
+
+  static unsigned int _idGenerator;
```

Comment:
Is there any reason this has to be global? I would much prefer it to be a part of the allocator itself. 
Also for thread-safety reasons, it would be a shame to have 2 different allocators break either other.

Reply:
None at all. Just a pattern I'm in the habit of. Can make debugging easier when you have multiple allocators and the ids are all unique. It needs to be atomic for thread safety in that case though, you are correct. Just removed the static.